### ISSUE TITLE
Test tserver restart scenario on an encryption at rest universe with a master that has no universe key

### ISSUE THREAD

**Issue by iSignal** [yugabyte/yugabyte-db#2727]
Jira Link: [DB-1526](https://yugabyte.atlassian.net/browse/DB-1526)

@rahuldesirazu brought up a scenario where a universe with encryption at rest enabled could get into a troubling scenario. This task is to track digging into that scenario in more detail.

If all masters lose their universe keys (perhaps by all restarting simultaneously), they are in then in a state that waits for yugaware or other external tool to input the latest universe key. In the meantime, if a tserver restarts, it would be unable to decrypt its local files and could get kicked out of the quorum by the master leader.

We need to make sure this process does not result in all 3 tservers losing data.

**Comment by rahuldesirazu** [comment]
Tested out this scenario on a 3 node RF=3 cluster. Performed a non-rolling master restart so that none of the masters had the latest key in memory, then restarted 2 tservers. A few things to note:
1. Tablets never went under RF =2. One set of peers gets deleted after raft considers them unavailable, but master has logic to prevent any tablet from going RF < 2, so we never end up triggering a delete on an under-replicated tablet.
2. Tablets stuck in bootstrap never get deleted. This is related to item https://github.com/yugabyte/yugabyte-db/issues/1260, but since some of the tablets are stuck in bootstrap, this process cannot be interrupted to delete the tablet. This is disadvantageous in clusters many nodes, since it won't let master pick another node to add the replica to.
cc: @bmatican

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-2727-analysis.md`):

# Analysis: yugabyte/yugabyte-db#2727 (DB-1526)

**Title:** Test tserver restart scenario on an encryption at rest universe with a master that has no universe key
**Kind:** scenario-analysis / enhancement (kind/enhancement, area/docdb, priority/medium), filed 2019-10-25, still OPEN.
**Ask:** if all masters lose their in-memory universe keys (simultaneous restart) and a tserver restarts before yugaware re-seeds the key, the tserver cannot decrypt its local files and "could get kicked out of the quorum by the master leader"; ensure this cannot result in all 3 tservers losing data.
**Review date:** 2026-08-21. Checkout: origin/master @ a27ae83fc10 (2026-08-19).

## Verdict
NOT-A-BUG — the depicted data-loss scenario is architecturally prevented on current master: a restarting tserver blocks pre-bootstrap until the key registry is retrievable (added by 20ac611182, 2.19.0.0+), raft never auto-evicts below 2 voters, and the masters-without-keys + tserver-restart flow is covered by `EncryptionTest.MasterSendKeys`. The enhancement's ask (dig in + ensure no data loss + test) is substantially done; issue can be closed or reduced to an optional extra test.

## Root cause
Not a defect; the 2019 hazard mechanism and its closure:

- **2019 premise still true by design (memory-only keys):** with `key_in_memory=true` (YBA flow), the latest universe key lives only in `EncryptionManager::universe_keys_`, an in-memory map (`src/yb/master/encryption_manager.h:91`, `src/yb/master/encryption_manager.cc:40`). Lookup is map-only: `GetKeyFromParams` returns `NotFound` when absent (`src/yb/master/encryption_manager.cc:246-255`). The persisted registry blob in sys_catalog cluster config (`EncryptionInfoPB.universe_key_registry_encoded`) is itself encrypted with that latest key (`src/yb/master/encryption_manager.cc:67-76`), so a fully-restarted master quorum cannot self-recover keys — intentional: keys must not persist next to the data they protect. Recovery = peer masters (async, retry-forever: `src/yb/master/catalog_manager.cc:1188,1799-1820`; `src/yb/client/universe_key_client.cc:46-90`) or external `AddUniverseKeys` from YBA/yb-admin (`src/yb/master/encryption_manager.cc:42-52`). With `key_in_memory=false` the key is read from a local file path (`src/yb/master/encryption_manager.cc:257-259`) and survives restarts — scenario does not apply.
- **Hazard closure (2023):** commit `20ac611182` "[#17075][#17078] Send full universe key registry to Tablet Server on Startup" (2023-05-05, first in 2.19.0.0) made `TabletServer::Init` block in an infinite linear-backoff retry loop calling `UniverseKeyClient::GetFullUniverseKeyRegistry` (`src/yb/tserver/tablet_server.cc:580-598`, backoff flags at :221-231) BEFORE constructing the encrypted envs (:600-604) and before `tablet_manager_->Init()` (:618). A tserver restarting while no master has the key therefore never attempts to decrypt local files, never bootstraps, never registers — it parks deterministically until the key is re-seeded. The 2019 failure mode (up-but-undecryptable replica flapping in the quorum) cannot occur.
- **Data-safety guards:** raft leaders evict an unreachable voter only after `follower_unavailable_considered_failed_sec` (900 s, `src/yb/consensus/consensus_queue.cc:84`) and only if `CountVoters(active_config) > 2` — "We never drop from 2 voters to 1 voter automatically" (`src/yb/consensus/consensus_queue.cc:599-610`). Eviction is a config change, not data deletion; the parked tserver's tablet services are not up, so no `DeleteTablet` can reach it; tombstone + remote-bootstrap re-add happen only after it returns (with keys). If 2 or 3 tservers are down, no config change can commit (majority unreachable) → no eviction at all. Master leader without keys simply omits the registry from heartbeats ("Leader master does not have universe key.", `src/yb/master/encryption_manager.cc:189-207`, called from `src/yb/master/catalog_manager_ext.cc:3609-3633`) — running tservers already hold the registry in memory and are unaffected.
- Worst case is therefore bounded **unavailability** of restarted tservers until the key is re-supplied (inherent to memory-only key custody), never loss of the encrypted data on disk. Matches the reporter's own 2019-11-20 test comment (tablets never dropped below 2 replicas).

## Trigger matrix
| # | Condition | Predicted (from source) | Observed (evidence) |
|---|---|---|---|
| 1 | `key_in_memory=true`, all masters restart simultaneously | keys absent; registry undecryptable until re-seeded | `encryption_manager.cc:246-255` NotFound; test asserts NOK at `encryption-test.cc:316` |
| 2 | tserver restarts while masters lack keys | blocks in retry loop pre-bootstrap; no decrypt attempt, no registration | `tablet_server.cc:580-598` `while (true)` before `:618` |
| 3 | blocked tserver unreachable >900 s, per-tablet leader alive | follower evicted only if voters > 2; never 2→1 | `consensus_queue.cc:599-610` |
| 4 | 2-3 of 3 tservers down | no majority → no config change → no eviction; unavailability only | raft commit rule; `MasterSendKeys` restarts all 3 and data verifies (`encryption-test.cc:321-328`) |
| 5 | key re-seeded via AddUniverseKeys | tservers finish Init; old+new keys delivered; data intact | `encryption-test.cc:299-329` (`MasterSendKeys`) |
| 6 | rolling master restart | peers re-seed each other | `catalog_manager.cc:1799-1820`; `encryption-test.cc:210-228` |
| 7 | `key_in_memory=false` (key_path) | masters self-recover from local file; scenario N/A | `encryption_manager.cc:257-259` |

All cells are source/test-evidence cells; no executable repro needed (see Reproduction).

## Reproduction
N/A (explanatory). This is a scenario-analysis enhancement with no failure signature; the dispatch accepted a source-architecture trace. The scenario is exercised deterministically by the existing external-mini-cluster test, runnable as:

```
./yb_build.sh release --cxx-test encryption-test --gtest_filter EncryptionTest.MasterSendKeys
```

which performs: write with key1 → rotate to key2 → write → stop ALL masters → restart them → assert no master has the key → seed only the latest key (YBA-style) → restart all 3 tservers → verify both keys delivered and all data readable (`src/yb/integration-tests/encryption-test.cc:299-329`). Companion coverage: `AllMastersRestart` (:196-208), `RollingMasterRestart` (:210-228), `MasterLeaderRestart` (:171-194), `ServerRestart` (:288-297).

## Versions tested
- Source trace on origin/master @ `a27ae83fc10` (2026-08-19): master affected = no (hazard closed).
- Hazard-closing commit `20ac611182` (2023-05-05) first released in 2.19.0.0 (tag `2.19.0.0-b144`+); all currently supported release lines (2024.x, 2025.x) contain it.
- 2019-era behavior (pre-2.19 startup without the blocking fetch) is what the issue describes; those lines are EOL.

## Proposed fix direction
No code fix warranted. Optional residual: `MasterSendKeys` seeds the key BEFORE restarting tservers; a test for the exact interleaving — restart tservers WHILE masters lack keys, assert they park (no crash, no bootstrap), then seed the key and assert recovery — would close the literal wording of the issue. That is a ~30-line variant of `MasterSendKeys` in `src/yb/integration-tests/encryption-test.cc` (shutdown tservers at :321 before the `AddUniverseKeyToAllMasters` at :317, restart, sleep past a few retry cycles, then seed and verify). Recommend closing #2727 as addressed by `20ac611182` / #17075 / #17078, optionally filing the small test addition.

## Test strategy
Existing target: `encryption-test` (`src/yb/integration-tests/encryption-test.cc`). Exercise via public entry points only (yb-admin `AddUniverseKeyToAllMasters` / `RotateUniverseKeyInMemory`, external mini-cluster restarts). One test per run:
`./yb_build.sh release --cxx-test encryption-test --gtest_filter EncryptionTest.MasterSendKeys`.

## Confidentiality check
Artifact contains only public-repo file paths, public commit hashes/tags, public issue/Jira IDs, and synthetic test details. No customer identifiers, PII, internal hostnames/URLs, or secrets.
